### PR TITLE
Updates builtin type Map so that it stores Maybes in its nodes.

### PR DESCRIPTION
**Pull request by In-Veritas**
## Summary of Changes

The following updates have been made to `builtins.bend`:

#### **1. `Maybe`**
- **New type definition:**
  ```bend
  type Maybe(T):
    Some {val: T}
    None
  ```

#### **2. `Map`**
- **Updated type definition:**
  ```bend
  type Map(T):
    Node {value: Maybe(T), ~left: Map(T), ~right: Map(T)}
    Leaf
  ```
  Now, every `Node` stores a `Maybe` type to explicitly represent the presence or absence of a value.

  ```
- **Updated/added functions:**
  - `Maybe/unwrap`: Retrieves the value inside Maybe/Some
  - `Map/set`: Inserts or updates a value in the map.
  - `Map/get`: Retrieves a value from the map, returning a `Maybe` type.
  - `Map/map`: Applies a function to each value in the map.
  - `Map/contains`: Checks whether a key exists in the map.
  - `Map/get_check`: Retrieves a value and performs additional validation.

---

## Why remove unreachable() from Map nodes?

   - You cannot see which elements are in the Map, nor can you perform a fold operation on the Map to, for example, return a list of values or map each value. Since the stored elements are not usable for anything, it is not possible to use this structure to determine whether a given element is present in the map or not.
   - The `unreachable()` function incorrectly ignores type check and can often cause bugs where the type isn't checked for an entire branch altogether.
   - When checking the results of a program, receiving `*` does not tell you whether you merely hit an unused node, or if another function called unreachable().

---

## Tests
  - Tests for the new functions have been added at Bend/tests/golden_tests/prelude

---

## Documentation update
Partially updates documentation adding a section for the Maybe type and new Map functions

### CODE REVIEW

File: tests/snapshots/run_file__unbound_wrap.bend.snap [hunk 1]
```diff
@@ -4,5 +4,5 @@ input_file: tests/golden_tests/run_file/unbound_wrap.bend
 ---
 [4m[1m[31mErrors:[0m
 [1mIn [4mtests/golden_tests/run_file/unbound_wrap.bend[0m[1m :[0m
-[1mIn definition '[4mmain[0m[1m':[0m
-  Reference to undefined function 'Maybe/wrap'
+[1mIn definition '[4mMaybe/bind[0m[1m':[0m
+  Unbound variable 'val.x'.
```
Comment (developedby) [inline tests/snapshots/run_file__unbound_wrap.bend.snap:8]: You broke this test. I'm guessing it defined its own Maybe and you removed that

**Review by developedby — CHANGES_REQUESTED** [review]
Why are you changing the `Map` type? You didn't even mention that in the PR.

Also please update the PR name to something more succint, its a bit confusing.

I'm not against having `Bool` in the prelude, but we don't really need it for anything builtin.

File: examples/parallel_and.bend [hunk 2]
```diff
@@ -1,8 +1,6 @@
-# This program allocates a tree with True at the leaves then parallel ANDs them.
-type Bool:
-  True
-  False
+type Bool = True | False
```
Comment (developedby) [inline examples/parallel_and.bend:1]: The comment was on the file not on the function. Go back to how it was

File: src/fun/builtins.bend [hunk 3]
```diff
@@ -170,62 +170,110 @@ def Tree/reverse(tree: Tree(T)) -> Tree(T):
...
-          0: (Map/Node unreachable (Map/set Map/Leaf (/ key 2) value) Map/Leaf)
-          _: (Map/Node unreachable Map/Leaf (Map/set Map/Leaf (/ key 2) value))
-        }
-        _: (Map/Node value Map/Leaf Map/Leaf)
-      }
-  }
+# MAYBE Impl
+
+type Maybe(T):
+  Some{val: T}
```
Comment (developedby) [inline src/fun/builtins.bend:176]: For the other types we've been writing the names in full, so
`Some { value: T }`

Also see the spacing

File: src/fun/builtins.bend [hunk 4]
```diff
@@ -170,62 +170,110 @@ def Tree/reverse(tree: Tree(T)) -> Tree(T):
...
+        (got, rest) = Map/get(map.left, (key / 2))
+        return(got, Map/Node(map.value, rest, map.right))
+      else:
+        (got, rest) = Map/get(map.right, (key / 2))
+        return(got, Map/Node(map.value, rest, map.left))
+    case Map/Leaf:
+      return (unreachable(), map)
+
+# Checks if a node has a value on a given key, returning Maybe/Some if it does, Maybe/None otherwise
+def Map/get_check (map: Map(T), key: u24) -> (Map(T), Maybe(T)):
```
Comment (developedby) [inline src/fun/builtins.bend:213]: `get_check` is not a great name. I prefer `get_checked` but it's also not great.

Comment (developedby) [inline src/fun/builtins.bend:213]: We've been using `(value, structure)` for these linear functions. See the type for `List/length` and `Map/get`

File: src/fun/builtins.bend [hunk 5]
```diff
@@ -170,62 +170,110 @@ def Tree/reverse(tree: Tree(T)) -> Tree(T):
...
+      if (0 == key):
+        return Map/Node(Maybe/Some(value), Map/Leaf, Map/Leaf)
+      elif ((key % 2) == 0):
+        return Map/Node(Maybe/None, Map/set(Map/Leaf, (key / 2), value), Map/Leaf)
+      else:
+        return Map/Node(Maybe/None, Map/Leaf, Map/set(Map/Leaf, (key / 2),value))
+
+
+# Checks if a Map contains a given key
+def Map/contains (map: Map(T), key: u24) -> (Map(T), u24):
```
Comment (developedby) [inline src/fun/builtins.bend:247]: Same thing here about the order of the return values

File: tests/golden_tests/compile_file/mismatched_ask_statements.bend [hunk 6]
```diff
@@ -9,4 +9,4 @@ def main:
         x <- wrap(0)
       case Bool/F:
         x = wrap(0)
-    return wrap(x)
+    return wrap(x)
```
Comment (developedby) [inline tests/golden_tests/compile_file/mismatched_ask_statements.bend:12]: Please don't remove the final line break

File: tests/golden_tests/compile_file_o_all/list_merge_sort.bend [hunk 7]
```diff
@@ -1,6 +1,7 @@
-type Bool = True | False
 type List_ = (Cons head tail) | Nil
 
+type Bool = True | False
```
Comment (developedby) [inline tests/golden_tests/compile_file_o_all/list_merge_sort.bend:3]: Unnecessary change

File: tests/golden_tests/compile_file_o_all/non_exhaustive_and.bend [hunk 8]
```diff
@@ -1,6 +1,5 @@
-type Bool = T | F
-
-Bool.and Bool/T Bool/T = Bool/T
-Bool.and Bool/F Bool/F = Bool/F
+type Bool = True | False
```
Comment (developedby) [inline tests/golden_tests/compile_file_o_all/non_exhaustive_and.bend:1]: Unnecessary change

File: tests/golden_tests/desugar_file/ask_branch.bend [hunk 9]
```diff
@@ -1,12 +1,9 @@
-type Bool:
-  T
-  F
-
+type Bool = True | False
```
Comment (developedby) [inline tests/golden_tests/desugar_file/ask_branch.bend:1]: Unnecessary change

File: tests/golden_tests/encode_pattern_match/and3.bend [hunk 10]
```diff
@@ -1,6 +1,6 @@
-type Bool = T | F
+type Bool = True | False
```
Comment (developedby) [inline tests/golden_tests/encode_pattern_match/and3.bend:1]: Unnecessary change

File: tests/golden_tests/encode_pattern_match/bool_tup.bend [hunk 11]
```diff
@@ -1,6 +1,5 @@
-type Bool = T | F
+type Bool = True | False
```
Comment (developedby) [inline tests/golden_tests/encode_pattern_match/bool_tup.bend:1]: Unnecessary change

File: tests/golden_tests/encode_pattern_match/definition_merge.bend [hunk 12]
```diff
@@ -1,5 +1,5 @@
-type Either = (Left value) | (Right value)
 type Bool = True | False
+type Either = (Left value) | (Right value)
```
Comment (developedby) [inline tests/golden_tests/encode_pattern_match/definition_merge.bend:2]: Unnecessary change

File: tests/golden_tests/encode_pattern_match/list_merge_sort.bend [hunk 13]
```diff
@@ -1,5 +1,5 @@
-type Bool = True | False
 type List_ = (Cons head tail) | Nil
+type Bool = True | False
```
Comment (developedby) [inline tests/golden_tests/encode_pattern_match/list_merge_sort.bend:2]: Unnecessary change

File: tests/golden_tests/encode_pattern_match/var_only.bend [hunk 14]
```diff
@@ -1,5 +1,3 @@
-type Bool = False | True
```
Comment (developedby) [inline tests/golden_tests/encode_pattern_match/var_only.bend:1]: I think this breaks the test

File: tests/golden_tests/mutual_recursion/odd_even.bend [hunk 15]
```diff
@@ -1,5 +1,4 @@
 type Bool = True | False
-
```
Comment (developedby) [inline tests/golden_tests/mutual_recursion/odd_even.bend:2]: Unnecessary change

File: tests/golden_tests/parse_file/imp_program.bend [hunk 16]
```diff
@@ -1,9 +1,6 @@
 type Point:
   Point { x, y }
-
-type Bool:
-  True
-  False
+type Bool = True | False
```
Comment (developedby) [inline tests/golden_tests/parse_file/imp_program.bend:3]: Unnecessary change

File: tests/snapshots/prelude__map_tests.bend.snap [hunk 17]
```diff
@@ -0,0 +1,6 @@
...
...e/None Map/Leaf λub (ub Map/Node/tag λvb (vb Maybe/Some/tag 13) Map/Leaf Map/Leaf)) λwb (wb Map/Node/tag Maybe/None λxb (xb Map/Node/tag Maybe/None λyb (yb Map/Node/tag Maybe/None Map/Leaf λzb (zb Map/Node/tag λac (ac Maybe/Some/tag 144) Map/Leaf Map/Leaf)) λbc (bc Map/Node/tag λcc (cc Maybe/Some/tag 77) Map/Leaf Map/Leaf)) λdc (dc Map/Node/tag λec (ec Maybe/Some/tag 42) λfc (fc Map/Node/tag Maybe/None Map/Leaf λgc (gc Map/Node/tag λhc (hc Maybe/Some/tag 255) Map/Leaf Map/Leaf)) λic (ic Map/Node/tag λjc (jc Maybe/Some/tag 100) Map/Leaf Map/Leaf)))), λkc (kc Map/Node/tag λlc (lc Maybe/Some/tag 1) λmc (mc Map/Node/tag Maybe/None λnc (nc Map/Node/tag Maybe/None λoc (oc Map/Node/tag Maybe/None Map/Leaf λpc (pc Map/Node/tag λqc (qc Maybe/Some/tag 66) Map/Leaf Map/Leaf)) λrc (rc Map/Node/tag λsc (sc Maybe/Some/tag 36) Map/Leaf Map/Leaf)) λtc (tc Map/Node/tag λuc (uc Maybe/Some/tag 9) Map/Leaf Map/Leaf)) λvc (vc Map/Node/tag λwc (wc Maybe/Some/tag 4) λxc (xc Map/Node/tag Maybe/None Map/Leaf λyc (yc Map/Node/tag λzc (zc Maybe/Some/tag 25) Map/Leaf Map/Leaf)) λad (ad Map/Node/tag λbd (bd Maybe/Some/tag 16) Map/Leaf λcd (cd Map/Node/tag λdd (dd Maybe/Some/tag 49) Map/Leaf Map/Leaf)))))))))))
```
Comment (developedby) [inline tests/snapshots/prelude__map_tests.bend.snap:6]: Don't write a test that returns a huge Map cause it's unreadable

File: tests/golden_tests/prelude/map_tests.bend [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+      (value, map) = Map/get(m, x)
+      return value
+    case 0:
+      return v
+
+#Checks if a generic map contains a given key, and if it does, applies a function to the value, otherwise it returns the map
+def test5(m: Map(T), key: u24) -> Map(T):
+  def self (x: T) -> T:
+    return x
+  match Map/contains(m, key):
```
Comment (developedby) [inline tests/golden_tests/prelude/map_tests.bend:46]: I don't know how this is compiling, but you have to use `switch` for matching on numbers.

File: tests/golden_tests/prelude/map_tests.bend [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+# Checks if a map contains a given key, and if it does, returns the assigned to the key, otherwise it returns the given value
+def test4(m: Map(T), x: u24, v: T) -> T:
+  match Map/contains(m, x):
+    case 1:
+      (value, map) = Map/get(m, x)
+      return value
+    case 0:
+      return v
+
+#Checks if a generic map contains a given key, and if it does, applies a function to the value, otherwise it returns the map
```
Comment (developedby) [inline tests/golden_tests/prelude/map_tests.bend:42]:
```suggestion
# Checks if a generic map contains a given key, and if it does, applies a function to the value, otherwise it returns the map
```

File: tests/golden_tests/prelude/map_tests.bend [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+    case 1:
+      return Map/map(m, key, self())
+    case 0:
+      return m
+
+#Checks if a u24 map contains a given key, and if it does, applies a function to the value, otherwise it returns the map
+def test6(m: Map(u24), key: u24) -> Map(u24):
+  def addtwo (x: u24) -> u24:
+    return (x + 2)
+  match Map/contains(m, key):
```
Comment (developedby) [inline tests/golden_tests/prelude/map_tests.bend:56]: same here

File: tests/golden_tests/prelude/map_tests.bend [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+
+
+def main():
+  m1 = {0: 42, 1: 7, 2: 13, 3: 255, 4: 100}  # all u24 numbers
+  m2 = {0: Bool/True, 1: Bool/False, 2: Bool/True, 3: Bool/False, 4: Bool/True}  # booleans
+  m3 = {0: Nat/Zero, 
+        1: Nat/Succ(Nat/Zero), 
+        2: Nat/Succ(Nat/Succ(Nat/Zero)), 
+        3: Nat/Succ(Nat/Succ(Nat/Succ(Nat/Zero)))}  # natural numbers
+  m4 = {0: 42,
```
Comment (developedby) [inline tests/golden_tests/prelude/map_tests.bend:70]: This is not well-typed so we don't need to test it (it'll output something impossible to read).

File: tests/golden_tests/prelude/map_tests.bend [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+  def addtwo (x: u24) -> u24:
+    return (x + 2)
+  match Map/contains(m, key):
+    case 1:
+      return Map/map(m, key, addtwo())
+    case 0:
+      return m
+
+
+def main():
```
Comment (developedby) [inline tests/golden_tests/prelude/map_tests.bend:63]:
```suggestion
def main() -> _:
```
Just to make sure the program is well-typed

File: tests/golden_tests/run_file/filter_bool_id.bend [hunk 23]
```diff
@@ -1,17 +1,15 @@
-type Bool:
-  T
-  F
+type Bool = True | False
```
Comment (developedby) [inline tests/golden_tests/run_file/filter_bool_id.bend:1]: Unnecessary change

File: tests/golden_tests/run_file/mixed_syntax.bend [hunk 24]
```diff
@@ -1,8 +1,5 @@
-type Bool:
-  True
-  False
-
 type MyTree = (node ~lft ~rgt) | (leaf val)
+type Bool = True | False
```
Comment (developedby) [inline tests/golden_tests/run_file/mixed_syntax.bend:2]: Unnecessary change

File: tests/golden_tests/run_file/unscoped_never_used.bend [hunk 25]
```diff
@@ -1,6 +1,5 @@
-type Bool = T | F
-
+type Bool = True | False
```
Comment (developedby) [inline tests/golden_tests/run_file/unscoped_never_used.bend:1]: Unnecessary change

File: tests/snapshots/run_file__nested_map_get.bend.snap [hunk 26]
```diff
@@ -3,7 +3,7 @@ source: tests/golden_tests.rs
 input_file: tests/golden_tests/run_file/nested_map_get.bend
 ---
 NumScott:
-43
+84
```
[the remark this answers is not in the document]
Comment (developedby) [reply tests/snapshots/run_file__nested_map_get.bend.snap:6]: This is still probably broken. Don't resolve unless you fix or have an explanation for this change

File: tests/snapshots/run_file__nested_map_set.bend.snap [hunk 27]
```diff
@@ -3,7 +3,7 @@ source: tests/golden_tests.rs
 input_file: tests/golden_tests/run_file/nested_map_set.bend
 ---
 NumScott:
-100
+*
```
[the remark this answers is not in the document]
Comment (developedby) [reply tests/snapshots/run_file__nested_map_set.bend.snap:6]: Same thing here

File: tests/golden_tests.rs [hunk 28]
```diff
@@ -471,6 +471,21 @@ fn io() {
...
 }
 
+/// Runs a file that uses the prelude.
+#[test]
+fn prelude() {
+  run_golden_test_dir(function_name!(), &|code, path| {
+    let _guard = RUN_MUTEX.lock().unwrap();
+    let book = parse_book_single_file(code, path)?;
+    let compile_opts = CompileOpts::default();
+    let diagnostics_cfg = DiagnosticsConfig::default();
```
Comment (developedby) [inline tests/golden_tests.rs:481]:
```suggestion
    let diagnostics_cfg = DiagnosticsConfig::new(Severity::Error, true);
```
There shouldn't be any warnings in the prelude library, so we can just error on them

**Review by developedby — CHANGES_REQUESTED** [review]

File: src/fun/builtins.bend (hunk 4 above)
Comment (In-Veritas) [reply src/fun/builtins.bend:213]: get_node?

Comment (developedby) [reply src/fun/builtins.bend:213]: `get` also gets the node

File: tests/golden_tests/prelude/map_contains_test.bend [hunk 29]
```diff
@@ -0,0 +1,3 @@
+def main():
+  m1 = {0: 42, 1: 7, 2: 13, 3: 255, 4: 100} 
+  return Map/contains(m1, 3)
```
Comment (developedby) [inline tests/golden_tests/prelude/map_contains_test.bend:3]: Please end all files with a new line

**Review by developedby — CHANGES_REQUESTED** [review]

File: src/fun/builtins.bend (hunk 4 above)
Comment (In-Veritas) [reply src/fun/builtins.bend]: get_wrapped?

File: docs/builtins.md [hunk 30]
```diff
@@ -173,7 +198,7 @@ type Map:
 **`Map`** represents a tree with values stored in the branches.
 It is meant to be used as an efficient map data structure with integer keys and O(log n) read and write operations.
 
-- **Node { value ~left ~right }**: Represents a map node with a `value` and `left` and `right` subtrees. Empty nodes have `*` stored in the `value` field.
+- **Node { value ~left ~right }**: Represents a map node with a `Maybe` and `left` and `right` subtrees. Empty nodes have `Maybe/None` stored in the `value` field, whislt non-empty nodes have `Maybe/Some` stored in the `value` field.
```
Comment (developedby) [inline docs/builtins.md:201]: Can you add the types here?

Comment (developedby) [inline docs/builtins.md:201]:
```suggestion
- **Node { value ~left ~right }**: Represents a map node with a `Maybe` and `left` and `right` subtrees. Empty nodes have `Maybe/None` stored in the `value` field, whilst non-empty nodes have `Maybe/Some` stored in the `value` field.
```

File: examples/example_fun.bend [hunk 31]
```diff
@@ -45,7 +45,6 @@ bad_nums : Any = (+ id 1)
 # Write new data types like this
 type Option = (Some val) | None
 type Bool = True | False
-
```
Comment (developedby) [inline examples/example_fun.bend:48]: Unnecessary change

File: tests/snapshots/prelude__builtin_Map_tests.bend.snap
Comment (developedby) [inline tests/snapshots/prelude__builtin_Map_tests.bend.snap:1]: Since you removed the test, please also remove the snapshot

File: tests/snapshots/prelude__get_values_from_map.bend.snap
Comment (developedby) [inline tests/snapshots/prelude__get_values_from_map.bend.snap:1]: The internal representation of Map is not really interesting for the tests, only the interfaces to it. When possible, try to return just a simple value, otherwise it's hard to understand if the snapshot is correct or not.

**Review by developedby — CHANGES_REQUESTED** [review]

File: tests/snapshots/prelude__map_tests.bend.snap
Comment (developedby) [inline tests/snapshots/prelude__map_tests.bend.snap:1]: I think this file can be removed, right?

**Review by developedby — CHANGES_REQUESTED** [review]
I'll trust the tests are correct

File: tests/snapshots/prelude__map_tests.bend.snap
Comment (In-Veritas) [reply tests/snapshots/prelude__map_tests.bend.snap:1]: Yes, it has been done

**Review by developedby — APPROVED** [review]
Looks good, good job

**Comment by developedby** [comment]
I think a new Rust version is messing with the checks, I'll fix it don't worry

**Comment by developedby** [comment]
Can you rebase and try again?
It should be fixed on the current main branch

**Review by developedby — APPROVED** [review]
